core-bundle/src/Menu/FrontendMenuBuilder.php

```diff
@@ -0,0 +1,366 @@
+            ) {
+                continue;
+            }
+
+            if (null === $href = $this->generateUri($page, $item)) {
+                continue;
+            }
+
+            $nextLevel = $level + 1;
+            $currentPageIsChild = $currentPage && ($currentPage->id === $page->id || \in_array($currentPage->id, $this->database->getChildRecords($page->id, 'tl_page'), false));
+            $displayChildren = !$options['showLevel']
+                               || $options['showLevel'] >= $nextLevel
+                               || (!$options['hardLimit'] && $currentPageIsChild);
+            $hasSubmenu = $hasSubpages && $displayChildren;
+
+            $this->populateMenuItem($item, $request, $page, $href, $hasSubmenu, $options);
+            $root->addChild($item);
+
+            // Allow modifying empty submenu nodes
+            if (!$hasSubpages) {
+                $menuEvent = new FrontendMenuEvent($this->factory, $item, (int) $page->id, $nextLevel, $options);
+                $this->dispatcher->dispatch($menuEvent);
+
+                continue;
+            }
+
+            $this->getMenu($item, (int) $page->id, $nextLevel, $host, $options);
+            $item->setDisplayChildren($displayChildren);
+        }
+
+        $menuEvent = new FrontendMenuEvent($this->factory, $root, $pid, $level, $options);
+        $this->dispatcher->dispatch($menuEvent);
+
+        return $root;
+    }
+
+    private function getPages(int $pid, array $options): array
+    {
+        // Custom page choice like, e.g., for the custom navigation module
+        if (0 === $pid && $options['pages']) {
+            return $this->findPagesByIds($options['pages']);
+        }
+
+        return $this->findPagesByPid($pid, (bool) $options['showHidden']);
+    }
+
+    /**
+     * @return array<array{page:PageModel,hasSubpages:bool}>
+     */
+    private function findPagesByIds(array $pageIds): array
+    {
+        // Get all active pages and also include root pages if the language is added to the URL (see #72)
+        $pages = $this->pageModelAdapter->findPublishedRegularByIds($pageIds, ['includeRoot' => true]);
+
+        if (null === $pages) {
+            return [];
+        }
+
+        return array_map(
+            static fn (PageModel $page): array => [
+                'page' => $page,
+                'hasSubpages' => false,
+            ],
+            $pages instanceof \Traversable ? iterator_to_array($pages) : $pages
+        );
+    }
+
+    /**
+     * @return array<array{page:PageModel,hasSubpages:bool}>
+     */
+    private function findPagesByPid(int $pid, bool $showHidden = false, bool $isSitemap = false): array
+    {
+        $time = Date::floorToMinute();
+        $blnBeUserLoggedIn = $this->tokenChecker->hasBackendUser() && $this->tokenChecker->isPreviewMode();
+        $unroutableTypes = $this->pageRegistry->getUnroutableTypes();
+
+        $pages = $this->connection
```

Comment:
This PR would finally allow us to improve this logic a bit. We execute this query for every single page and it can never be cached. However, if we were to restructure the logic a bit, we could maybe improve this.
Given the following structure:

- 1
    - 2
        - 21
        - 22
        - 23
    - 3
        - 31
        - 32
        - 33

Currently, if we render the navigation for ID 1, it will execute this query for every single page. However, e.g. for ID 2 and 3 we already know the respective 3 subpages. So instead of requesting that query for every single one of them, maybe they could be loaded in a single one?

Reply:
Mhhh, I am unsure about this.

First, the query actually does not return the IDs of the next level.
Second, recursive queries are only possible from MySQL 8+; it looks like: https://stackoverflow.com/a/33737203

All I can offer is that we cache the information whether the page has subpages or not. This will save some resources for re-querying empty pages. See https://github.com/richardhj/contao/commit/6587efc470c02bc2cfa4969619378ddd8cb7e111 for this purpose.
However, this comes with problems, e.g., different navigation modules on the same page differ in the $showHidden option.

But maybe this is all about my inability to write complex MySQL queries 😄

Reply:
Maybe @ausi has an idea? 😊

Reply:
In my experience the "with" or "connect by prior" functionality of DBMS are most of the time more suited for getting a sorted list (in "tree form") than to save performance. And as @richardhj already stated, it's not supported in older MySQL versions.

If it's just about saving requests, it would probably be faster and much simpler to just get a projection of all records - in this case a `SELECT id, pid FROM tl_page WHERE <published, not hidden, …>` (once) to build a lookup table.